### PR TITLE
AN-109087: Add time/date encoder in kdbgo

### PR DESCRIPTION
**Pull request by yznima**
1. Support encoding Time & Date.
2. Added more tests with boundaries for time & date
3. Added decoder for time. (Date will be supported as a followup PR)

### CODE REVIEW

**Comment by jkessler93** [comment]
Can you add a description?

File: encode.go [hunk 1]
```diff
@@ -11,74 +11,92 @@ import (
...
 
 // TODO: Handle all the errors
-func writeData(dbuf *bytes.Buffer, order binary.ByteOrder, data *K) error {
-	binary.Write(dbuf, order, data.Type)
-	if K0 <= data.Type && data.Type <= KT { // All Lists
-		binary.Write(dbuf, order, data.Attr)
-		binary.Write(dbuf, order, int32(reflect.ValueOf(data.Data).Len()))
-	} else if data.Type == XT { // Table
-		binary.Write(dbuf, order, data.Attr)
+func write(buf *bytes.Buffer, order binary.ByteOrder, k *K) error {
```
Comment (jkessler93) [inline encode.go]: I think in transit we call this qStruct, should we be consistent here?

File: encode.go [hunk 2]
```diff
@@ -11,74 +11,92 @@ import (
...
 			}
 		}
-	case -KB, -UU, -KG, -KH, -KI, -KJ, -KE, -KF, -KC, -KD, -KT, // TODO: case: -KM , -KZ, -KN, -KU, -KV
-		KB, UU, KG, KH, KI, KJ, KE, KF, KM, KD, KZ, KN, KU, KV, KT:
+	case -KB, -UU, -KG, -KH, -KI, -KJ, -KE, -KF, -KC, -KM , -KZ, -KN, -KU, -KV,
+		+KB, +UU, +KG, +KH, +KI, +KJ, +KE, +KF, +KM, +KZ, +KN, +KU, +KV:
 		// Scalar & Vector of boolean, int variants, float variant, and bytes.
-		// Note: UUID is backed by byte array of length 16
-		binary.Write(dbuf, order, data.Data)
+		// Note: UUID is back by byte array of length 16
```
Comment (jkessler93) [inline encode.go]: (nit) backed

File: encode.go [hunk 3]
```diff
@@ -11,74 +11,92 @@ import (
...
+			buf.WriteString(toSend)
+			binary.Write(buf, order, byte(0)) // Null terminator
 		}
 	case -KP: // Scalar timestamp
-		binary.Write(dbuf, order, data.Data.(time.Time).Sub(qEpoch))
+		binary.Write(buf, order, k.Data.(time.Time).Sub(qEpoch))
 	case KP: // Vector timestamp
-		for _, toSend := range data.Data.([]time.Time) {
-			binary.Write(dbuf, order, toSend.Sub(qEpoch))
+		for _, toSend := range k.Data.([]time.Time) {
```
Comment (jkessler93) [inline encode.go]: should we rename `toSend` to something like `timeVal`?

File: encode.go [hunk 4]
```diff
@@ -11,74 +11,92 @@ import (
...
+		binary.Write(buf, order, int32(days))
+	case KD: // Vector date
+		for _, date := range k.Data.([]time.Time) {
+			days := date.Sub(qEpoch) / (time.Hour * 24)
+			binary.Write(buf, order, int32(days))
+		}
+	case -KT: // Scalar time
+		t := k.Data.(time.Time)
+		nanos := t.Hour()*int(time.Hour) + t.Minute()*int(time.Minute) + t.Second()*int(time.Second) + t.Nanosecond()
+		binary.Write(buf, order, int32(nanos/int(time.Millisecond)))
```
Comment (jkessler93) [inline encode.go]: Can we name `int32(nanos/int(time.Millisecond))` to a variable `millis`?

**Review by jkessler93 — APPROVED** [review]
🎉

File: encode.go (hunk 2 above)
Comment (yznima) [reply encode.go]: I don't know how this happened. Will update

File: encode.go (hunk 3 above)
Comment (yznima) [reply encode.go]: Its actually timestamp. What do you think about `ts` or `timestamp`?

**Comment by yznima** [comment]
@appian/datalayer-ha Could you review?

File: encode.go (hunk 1 above)
Comment (yznima) [reply encode.go]: Since they are in two different repositories, I think its ok to leave it as `k`

**Comment by vy-ton** [comment]
Starting review

File: encode.go (hunk 1 above)
Comment (vy-ton) [inline encode.go]: Should we avoid change function/param names?

File: encode.go [hunk 5]
```diff
@@ -11,74 +11,92 @@ import (
...
-func writeData(dbuf *bytes.Buffer, order binary.ByteOrder, data *K) error {
-	binary.Write(dbuf, order, data.Type)
-	if K0 <= data.Type && data.Type <= KT { // All Lists
-		binary.Write(dbuf, order, data.Attr)
-		binary.Write(dbuf, order, int32(reflect.ValueOf(data.Data).Len()))
-	} else if data.Type == XT { // Table
-		binary.Write(dbuf, order, data.Attr)
+func write(buf *bytes.Buffer, order binary.ByteOrder, k *K) error {
+	binary.Write(buf, order, k.Type)
+	if K0 <= k.Type && k.Type <= KT { // All Lists
```
Comment (vy-ton) [inline encode.go]: `// All vectors` to be consistent with q terminology?

Comment (vy-ton) [inline encode.go]: Maybe a comment above `// Write q attributes (`s,`u,`g, etc...)`

File: encode.go [hunk 6]
```diff
@@ -11,74 +11,92 @@ import (
 )
 
 // TODO: Handle all the errors
```
Comment (vy-ton) [inline encode.go]: Maybe a comment here `Negative constants are scalars, positive are vectors`? Then the comments down below could just say the type instead of repeated scalar and vector

File: encode.go [hunk 7]
```diff
@@ -11,74 +11,92 @@ import (
...
+			days := date.Sub(qEpoch) / (time.Hour * 24)
+			binary.Write(buf, order, int32(days))
+		}
+	case -KT: // Scalar time
+		t := k.Data.(time.Time)
+		nanos := t.Hour()*int(time.Hour) + t.Minute()*int(time.Minute) + t.Second()*int(time.Second) + t.Nanosecond()
+		binary.Write(buf, order, int32(nanos/int(time.Millisecond)))
+	case KT: // Vector time
+		for _, t := range k.Data.([]time.Time) {
+			nanos := t.Hour()*int(time.Hour) + t.Minute()*int(time.Minute) + t.Second()*int(time.Second) + t.Nanosecond()
```
Comment (vy-ton) [inline encode.go]: Is the untyped `int` safe here?

File: struct.go [hunk 8]
```diff
@@ -330,25 +352,17 @@ func (m Month) String() string {
...
 // Second hh:mm:ss
 type Second time.Time
 
 func (s Second) String() string {
-	time := time.Time(s)
-	return fmt.Sprintf("%02v:%02v:%02v", time.Hour(), time.Minute(), time.Second())
-}
-
-// Time hh:mm:ss.SSS
-type Time time.Time
```
Comment (vy-ton) [inline struct.go:347]: I assume they created the alias to have additional receiver functions?

**Review by vy-ton — APPROVED** [review]
LGTM! Just a few suggestions on additional comments

File: encode.go (hunk 1 above)
Comment (yznima) [reply encode.go]: Sure I will revert this

File: encode.go (hunk 7 above)
Comment (yznima) [reply encode.go]: It is fine since all of them are within 32 bit range but I will make it int64 as its more accurate

File: struct.go (hunk 8 above)
Comment (yznima) [reply struct.go:347]: Yes the only additional type was the string method. Since I introduced the Time function, it was clashing with this, hence I removed this as I don't think it adds much value

File: encode.go (hunk 7 above)
Comment (yznima) [reply encode.go]: Actually you are right. It wasn't safe. Updated to use `int64`

**Comment by yznima** [comment]
@appian/datalayer-ha This is ready to merge
